ratis-logservice/src/main/java/org/apache/ratis/logservice/server/LogStateMachine.java

```diff
@@ -452,29 +503,166 @@ public void close() {
+            }
+            return true;
+          };
+        archiveFuture = executorService.submit(callable);
+      }
+      return CompletableFuture.completedFuture(
+          Message.valueOf(LogServiceProtoUtil.toArchiveLogReplyProto(t).toByteString()));
+    } catch (Throwable e) {
+      return CompletableFuture.completedFuture(
+          Message.valueOf(LogServiceProtoUtil.toArchiveLogReplyProto(e).toByteString()));
+    }
+  }
+
+  private void sendArchiveLogrequestToNewLeader(long recordId, LogName logName, String location)
+      throws IOException {
+    getClient().sendReadOnly(
+        () -> LogServiceProtoUtil.toArchiveLogRequestProto(logName, location, recordId, isArchival)
+            .toByteString());
+  }
+
+
+  private void commit(ArchiveLogWriter writer, LogName logName, String location)
+      throws IOException {
+    writer.rollWriter();
+    updateArchivingInfo(writer.getLastWrittenRecordId(), logName, location, isArchival);
+  }
+
+  private void updateArchivingInfo(long recordId, LogName logName, String location,
+      boolean isArchival)
+      throws IOException {
+    RaftClientReply archiveLogReply = getClient().send(
+        () -> LogServiceProtoUtil.toArchiveLogRequestProto(logName, location, recordId, isArchival)
+            .toByteString());
+    LogServiceProtos.ArchiveLogReplyProto message=LogServiceProtos.ArchiveLogReplyProto
+        .parseFrom(archiveLogReply.getMessage().getContent());
+    if (message.hasException()) {
+      throw new IOException(message.getException().getErrorMsg());
+    }
+  }
+
+  private void sendChangeStateRequest(State state) throws IOException {
+    if (isArchival) {
+      getClient().send(
+          () -> LogServiceProtoUtil.toChangeStateRequestProto(LogName.of("Dummy"), state)
+              .toByteString());
+    }
+
+  }
+
+  private RaftClient getClient() throws IOException {
+    if (client == null) {
+      try {
+        RaftServer raftServer = server.get();
+        client = RaftClient.newBuilder().setRaftGroup(getGroupFromGroupId(raftServer, groupId))
+            .setClientId(ClientId.randomId())
+            .setProperties(raftServer.getProperties()).build();
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+    }
+    return client;
+  }
+
+  private RaftGroup getGroupFromGroupId(RaftServer raftServer, RaftGroupId raftGroupId)
+      throws IOException {
+    List<RaftGroup> x = StreamSupport.stream(raftServer.getGroups().spliterator(), false)
+        .filter(group -> group.getGroupId().equals(raftGroupId)).collect(Collectors.toList());
+    if (x.size() == 1) {
+      return x.get(0);
+    } else {
+      throw new GroupMismatchException(x.size() + " are group found for group id:" + raftGroupId);
+    }
+  }
+
+  @Override public void notifyNotLeader(Collection<TransactionContext> pendingEntries)
+      throws IOException {
+    isNoMoreLeader = true;
```

Comment:
Is it possible to "interrupt" the transactions which are pending? Could we keep a `Map<TransactionContext,Callable>` where the callable is doing the archiving? That might be more clean?